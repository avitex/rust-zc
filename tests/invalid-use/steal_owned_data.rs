use zc::{Dependant, Zc};
use once_cell::sync::OnceCell;

static THIEF: OnceCell<&'static [u8]> = OnceCell::new();

#[derive(Dependant)]
pub struct MyStruct<'a>(&'a [u8]);

fn steal_owned_data(bytes: &'static [u8]) -> MyStruct<'static> {
    THIEF.get_or_init(|| bytes);
    MyStruct(bytes)
}

fn main() {
    let owner = vec![1, 2, 3];

    let _ = Zc::new(owner, steal_owned_data);

    // should not work
    assert_eq!(THIEF.get().unwrap(), &[1, 2, 3]);
}
