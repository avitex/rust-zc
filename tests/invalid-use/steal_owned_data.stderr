error: implementation of `zc::private::Construct` is not general enough
  --> $DIR/steal_owned_data.rs:17:13
   |
17 |       let _ = Zc::new(owner, steal_owned_data);
   |               ^^^^^^^ doesn't satisfy where-clause
   |
  ::: $WORKSPACE/src/private.rs
   |
   | / pub unsafe trait Construct<'o, O>: Sized
13 | | where
14 | |     O: ?Sized,
15 | | {
...  |
18 | |     unsafe fn construct(self, owned: &'o O) -> Self::Dependant;
19 | | }
   | |_- trait `zc::private::Construct` defined here
   |
  ::: $WORKSPACE/src/lib.rs
   |
   | /     pub fn new<C>(owner: O, constructor: C) -> Self
74 | |     where
75 | |         C: for<'o> Construct<'o, <O::Storage as Deref>::Target, Dependant = D>,
76 | |     {
...  |
81 | |         Self { storage, value }
82 | |     }
   | |_____- due to a where-clause on `Zc::<O, D>::new`...
   |
   = note: ...`fn(&'static [u8]) -> MyStruct<'static> {steal_owned_data}` must implement `zc::private::Construct<'0, [u8]>`, for any lifetime `'0`...
   = note: ...but `fn(&'static [u8]) -> MyStruct<'static> {steal_owned_data}` actually implements `zc::private::Construct<'1, [u8]>`, for some specific lifetime `'1`
